SuperteamDAO/earn
# Fix | Editable Slugs on UI with auto generate

## What does this PR do?
This PR fixes https://github.com/SuperteamDAO/earn/issues/479

## Where should the reviewer start?
The implementation can be divided into 3 sections:
- UI: Auto generate slug based on title (ListingBasic.tsx)
- UI: Check slug for validity (isSlugAvailable.ts)
- Server: Validation API and changes to create listing API (create.ts and slug.ts)

## How should this be manually tested?
1. Create a listing as normal and see that the slug gets generated automatically based on the title
2. Create a listing as normal and edit the slug manually
3. Create a new listing with the same slug as before and check that the error shows
4. Create a new listing, edit the slug to use invalid chars (non alphabet and non number), check that the error shows

## Any background context you want to provide?
I identified a few ways in which the entire form related logic (input, validation and submission) can be simplified using reusable components, I'm happy to discuss these. 
Additionally, few assumptions made during the process: 
1. the API call to check slug validity is made on blur, to reduce the number of calls in total
2. When the slug is edited manually, and then the listing title is changed, the slug is auto updated based on the new title, this can be changed trivially such that if the slug is manually edited once, it stays that way even when the title changes

## What are the relevant issues?
https://github.com/SuperteamDAO/earn/issues/479


## Screenshots (if appropriate)
![image](https://github.com/SuperteamDAO/earn/assets/31418899/50df4aff-0fae-484e-8224-01c0878f6912)
![image](https://github.com/SuperteamDAO/earn/assets/31418899/d34f0806-84d4-4ce1-8b1c-391ad68a8703)
![image](https://github.com/SuperteamDAO/earn/assets/31418899/54ffe75d-a3ae-4aa2-8928-4b87f66ae4a3)
